scripts/gather-licenses.ts

```diff
@@ -0,0 +1,218 @@
+      .sort()
+      .map(async filename => ({
+        filename,
+        content: await fs.readFile(path.join(packagePath, filename), 'utf8')
+      })));
+
+  return {
+    ...packageJSON,
+    path: packagePath,
+    licenseFiles
+  };
+}
+
+// Recursively look up dependencies of a package at `startPath`, and return
+// the package information for all of these packages.
+// Each package name + version combination is only reported once.
+export async function gatherLicenses(startPath: string): Promise<Package[]> {
+  const packages = new Map<string, Package>();
+
+  packages.set(startPath, await getPackageInfo(startPath));
+  for (const pkg of packages.values()) {
+    const require = createRequire(path.resolve(pkg.path, 'index.js'));
+    for (const depsKey of ['dependencies', 'optionalDependencies'] as const) {
+      for (const dep of Object.keys(pkg[depsKey] ?? {})) {
+        let resolved;
+        try {
+          resolved = require.resolve(dep);
+          if (!path.isAbsolute(resolved)) {
+            continue; // Node.js builtin
+          }
+        } catch (err) {
+          resolved = await findUp(async directory => {
+            const candidate = path.join(directory, 'node_modules', dep, 'package.json');
+            return await findUp.exists(candidate) ? candidate : undefined;
+          }, { cwd: path.resolve(pkg.path) });
+        }
+        if (!resolved) {
+          if (depsKey === 'optionalDependencies') {
+            continue;
+          }
+          throw new Error(`Could not resolve ${dep} require from ${pkg.path}`);
+        }
+        const pkgJson = await pkgUp({ cwd: path.dirname(resolved) });
+        if (!pkgJson) {
+          throw new Error(`Could not find package.json for ${dep} required from ${pkg.path}`);
+        }
+        const pkgBase = path.relative(process.cwd(), await fs.realpath(path.dirname(pkgJson)));
+        if (!packages.has(pkgBase)) {
+          packages.set(pkgBase, await getPackageInfo(pkgBase));
+        }
+      }
+    }
+  }
+
+  // Deduplicate by package name and version.
+  const packageByNameAndVersion = new Map<string, Package>();
+  for (const pkg of packages.values()) {
+    const nameAndVersion = id(pkg);
+    const existing = packageByNameAndVersion.get(nameAndVersion);
+    if (existing) {
+      if (!pkgEqual(pkg, existing)) {
+        throw new Error(`Two entries for the same package name and version differ! (${JSON.stringify([pkg, existing])})`);
+      }
+    } else {
+      packageByNameAndVersion.set(nameAndVersion, pkg);
+    }
+  }
+  return [...packageByNameAndVersion.values()].sort((a, b) => a.name.localeCompare(b.name));
+}
+
+// Helper function for making sure that if we do encounter a package twice,
+// the package information roughly match.
+function pkgEqual(a: Package, b: Package): boolean {
+  for (const key of relevantPackageKeys) {
+    if (!util.isDeepStrictEqual(a[key], b[key])) {
+      console.log(a.name, b.name, key)
```

Comment:
Hmm, not sure if this was left there intentionally because the error message seems to contain both package files anyway 🤔

Reply:
Yes! It was left here unintentionally, thanks for mentioning it. :)